drivers/spi/spi_sam0.c

```diff
@@ -219,6 +219,13 @@ static void spi_sam0_fast_rx(SercomSpi *regs, const struct spi_buf *rx_buf)
 	regs->DATA.reg = 0;
 	len--;
 
+	/* Ensure the data register has shifted to the shift register before
+	 * continuing.	Later writes are synchronised by waiting for the receive
+	 * to complete.
+	 */
+	while (!regs->INTFLAG.bit.DRE) {
```

Comment:
TBH, this while loops, without any limitation, scare me.

Reply:
This is an idiomatic way of doing embedded device drivers though - by writing to a register and then spinning until the peripheral acknowledges.  It's used more than 700 times in the drivers/ directory.

I'd agree if the operation could timeout in some way, or if the operation could fail in someway, but I don't know of a failure mode here that would be better handled by timing out and returning an error instead of waiting for the watchdog.

Reply:
If you can guaranty that after a number n of loops this register _will be set_ then use such limit.

About timeout, that's something which may be introduced not so far away. See #24511 (though it will require some work to get something usable)

Reply:
I'm reluctant to change this without consensus on how this should be done across the whole of Zephyr.  The issues are that this PR follows the style of the rest of Zephyr, this error is rare, is not actionable by the developer, and leads to more bugs than it fixes.  As a secondary issue, it's in the performance path of this function.

re: rare, I don't know of any normal error conditions that could cause this to loop forever.  This is an on-chip peripheral so you're in the same probability class as random memory corruption.

re: actionable, what should the developer do if they get this error?  You can't recover with reinitialising the peripheral, which in Zephyr is done at boot time.  It's better to isolate this to a non-critical thread or wait for the watchdog to trigger.  And in a safe design, you already need to understand and properly handle what happens when the watchdog triggers.

re: bugs, I've lost count of the number of bugs I've seen due to a too short or buggy timeout.  In fact, here's one I fixed in OpenOCD just recently: http://openocd.zylin.com/#/c/5706/

Does Zephyr have an error philosophy?

Reply:
I did not request changes or block it, but I can not approve, it does not look right to me how this driver is designed.

Reply:
> I'd agree if the operation could timeout in some way, or if the operation could fail in someway, but I don't know of a failure mode here that would be better handled by timing out and returning an error instead of waiting for the watchdog.

If an interrupt does not occur here, there must be something wrong within the chip and/or the software system supporting it, so the system is already compromised and deserves to be nuked by the watchdog.

Adding timeout here will only provide a false sense of security.